### PR TITLE
Fix tempfile issues

### LINKED ISSUE #4330: No session_* subfolder if temp_dir is specified
Currently if `temp_dir` is specified, the sockets and logs are directly stored there, and not in a `session_${DATE}_${PROCESS_ID}` subfolder. This may lead to logs being mixed from multiple runs and/or collisions for the socket files if the `temp_dir` is hardcoded in a script which is run multiple times.

cc @nileshtrip

### PR DESCRIPTION
**Pull request by suquark**
## What do these changes do?

This PR addresses some crucial tempfile issues.

## Related issue number

Closes #4330 #4593 #4594

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

## TODO list:

- [ ] ~~Fix autoscaler (https://github.com/ray-project/ray/pull/4331#issuecomment-472624354)~~ (see https://github.com/ray-project/ray/issues/4737)
- [x] Add more unittests
- [x] Update related document

### CODE REVIEW

**Comment by suquark** [comment]
Ready for review

File: python/ray/tests/test_tempfile.py [hunk 1]
```diff
@@ -128,3 +122,20 @@ def test_raylet_tempfiles():
...
+    os.chmod("/tmp/ray", 0o000)
+    ray.init(num_cpus=1)
+    session_dir = ray.worker._global_node.get_session_dir_path()
+    assert os.path.exists(session_dir), "Specified socket path not found."
+    ray.shutdown()
+
+
+def test_session_dir_uniqueness():
+    session_dirs = set()
+    for _ in range(50):
```
Comment (robertnishihara) [inline python/ray/tests/test_tempfile.py]: 50 is too large. That could add around 50 seconds to the test. How about 2 or 3?

**Review by robertnishihara — COMMENTED** [review]
@suquark nice work, I tried it out and this seems to work great.

Do we have a test to make sure that the `--temp-dir` flag to `ray start` does the right thing? And in particular, that when that is used, `/tmp/ray` is not created?

**Review by robertnishihara — APPROVED** [review]
Nice work!

**Comment by mitar** [comment]
Has this addressed #4641?

**Comment by robertnishihara** [comment]
I think so, thanks!

**Comment by guoyuhong** [comment]
@suquark It looks like this PR introduces some CI failures to master and I can also observe this test failure in my other PRs. Could you take a look? https://travis-ci.com/ray-project/ray/jobs/198094058
image

**Comment by suquark** [comment]
@guoyuhong It's a bit strange that the temp dir wasn't deleted. Let me investigate it.

**Pull request merged**